sprint 2 - step 1

ArtemKorneyasev opened the pull request:
- Add React.Context for experimental puprose
- Add request for creating order
- Refactoring

Ready for soft review

ArtemKorneyasev committed:
sprint 2 - step 1

- Add React.Context for experimental puprose
- Add request for create order
- Refactoring

irinamozes commented on src/components/app/app.jsx:
```diff
@@ -23,7 +24,7 @@ const App = () => {
 			} catch (error) {
 				setState({ ...state, hasError: true });
 			}
```
При такой структуре конструкции try-catch, обнаружиться в блоке catch и обработаться могут только ошибки сети, или когда сервер будет недоступен. В случае, когда сервер вернёт статус ошибки (response.status>=300), эта ситуация в блоке catch никак не обнаружится. Поэтому, после выполнения запроса методом fetch,
сразу должна быть проверка, какой статус вернул сервер, и дальнейшие действия в соответствии с результатом этой проверки (выброс исключения, если сервер вернул статус ошибки, которое и обнаружится в catch). Изучите подобный пример, разобранный здесь:
https://developer.mozilla.org/ru/docs/Web/API/Fetch_API/Using_Fetch#проверка_успешности_запроса раздел "Проверка успешности запроса" (лучше прочитать всю статью) и сделайте по аналогии.

В заданиях ранее говорилось, что адрес запроса к серверу должен быть помещён в переменную. Эта переменная и должна использоваться как аргумент fetch.

Выполните вышеизложенные 2 замечания и для запроса номера заказа.

ArtemKorneyasev replied:
Поправил

irinamozes commented:
Проект очень хороший.

Правильно сделана вёрстка.

Что нужно подкорректировать.

1. Папку services нужно вынести из папки components и поместить в src, так как там нет компонент.

2. Вы передаёте массив ingredients как значение контекста в компоненте Provider. Объекты, которые передаются в контексте не должны передаваться и пробрасываться из компоненты в компоненту как пропсы (в этом и есть смысл контекста). Компоненты, котором нужны эти объекты, должны быть подписаны на контекст и получать
их из контекста. Подпишите компоненту BurgerIngredients на контекст. Проверьте, нет ли у Вас ещё компонент, которые вместо контекста, нужные объекты получают из пропсов.

3. Структура конструкций try-catch в запросах к серверу должна быть преобразована (подробный комментарий в app.js).

4. Адрес запроса к серверу должен быть помещён в переменную (подробный комментарий в app.js).

5. Поскольку в описании задания сказано:"Проверьте корректность расчётов [стоимости]: генерируйте разные ингредиенты в компоненте BurgerConstructor.", должна быть другая функциональность сбора бургера в конструкторе, более приближенная к варианту, который будет на втором этапе.

При загрузке сайта окно конструктора должно быть пустым.

Алгоритм по сборке бургера в конструкторе должен быть следующим:

Все ингредиенты в бургер добавляются (пока, на этом этапе) кликом по ингредиенту в окне ингредиентов(вместе с открытием модального окна ингредиента).

Запрос к серверу за номером заказа нельзя отправить пока не добавлена булка.

При добавлении булки в бургер, она сразу добавляется  и в верх, и в низ бургера.

При добавлении другой булки в бургер, она заменяет существующую булку, если таковая есть,
при этом сразу и верх, и низ бургера, то есть в бургере может быть всего 2 булки, вверху и внизу.

Стоимость рассчитывается как сумма всех ингредиентов добавленных в конструктор. В стоимость должны быть включены две булки - верх и низ.
Стоимость рассчитывается одновременно с добавлением ингредиента, новое значение сразу же выводится на экран.

Можно осуществить и удаление ингредиента из бургера при клике на значок корзинки и пересчёт цены, но этого в задании пока не требуется в обязательном порядке.

Для осуществления этого алгоритма лучше воспользоваться хуком useReducer, это поможет с освоением Redux на втором этапе.


Что я бы рекомендовала также переделать.

1.По смыслу показа модального окна в портале, а также из описания 2-го этапа 1-го спринта следует, что в проекте должно быть одно модальное окно (Modal), которое должно закрываться по крестику, являющемуся элементом Modal, по оверлею (ModalOverley) - также дочернему элементу Modal, и Esc.
Окно Modal, демонстрирующееся в портале, может наполняться каким угодно содержимым. В проекте это содержимое должно олицетворяться компонентами IngredientDetails и OrderDetails. 
Рекомендовала бы следующий алгоритм работы с порталом (так всё равно будет необходимо сделать в 3-м спринте, когда Вам нужно будет осуществить роутинг в приложении).
Портал нужно организовать в компоненте Modal и рендерить в ней компоненты IngredientDetails, OrderDetails как children, в портале.
Таким образом, можно саму компоненту Modal рендерить в App (или другой главной родительской компоненте) 2 раза (каждый экземпляр Modal должен показываться, или не показываться в зависимости от каких-то условий), и внутри экземпляров Modal как атрибут children рендерить IngredientDetails или OrderDetails.
Также в портале Modal нужно рендерить и ModalOverlay, но, поскольку его нужно рендерить всегда, его не надо рендерить как children, а можно просто как соседний элемент к children.
Функцию закрытия Modal по крестику нужно определить в App ( чтобы не определять её два раза) и передавать  как проп обоим экземплярам Modal.
Теория портала в лекциях находится здесь: https://praktikum.yandex.ru/learn/react/courses/b512c4ac-8797-4b96-935a-6b2865f27c9a/sprints/6021/topics/3daef531-37a0-4521-ac9c-dac879418335/lessons/f60fe57f-6c77-446c-966d-1fefe0413a9f/

Commit:
refactoring after soft review

- fetch requests improvements
- burger-ingredients wrapped by React.Context
- services directory moved to src/

ArtemKorneyasev committed:
try to implement useReducer part#1

ArtemKorneyasev committed:
refactoring after soft review

- implement useReducer hook
- use global AppContext
- change app logic

ArtemKorneyasev commented:
@irinamozes, спасибо за фидбэк. Извиняюсь за долгий рефакторинг, не все сразу мне далось, но все замечания в последних коммитах я исправил. Просьба проверить.

ArtemKorneyasev committed:
action 'ingredientError' now keep error description

ArtemKorneyasev committed:
small fix in appReducer, add payload to action 'ingredientsError'

ArtemKorneyasev committed:
preparing for redux usage

irinamozes commented:
Проект стал почти отличным.

Теперь правильно показывается модальное окно в портале, без дублирования функций его открытия и закрытия.

Подготовлен переход к Redux.

Что нужно подкорректировать.

1. После того как была попытка отправить заказ без булок, при повторной отправке, когда булки уже добавлены, вместо окна заказа появляется окно с сообщением, что нет булок (см. снимок). Исправьте это, пожалуйста, ко второму этапу.
![Снимок экрана в 2021-07-14 19-54-03](https://user-images.githubusercontent.com/19558891/125647049-5e8511b5-d742-4a2c-8be7-6a36d1833af8.png)

Задание принято.

Желаю удачи!

P.S. Извините, пожалуйста, за долгое ожидание проверки - вчера в течении всего дня были проблемы с сервером для студенческих проектов
и сегодня у меня были перебои с интернетом.

ArtemKorneyasev committed:
sprint-2/step-1 done
